File: src/main/java/com/bluehabit/budgetku/component/role/service/RoleService.java
/*
 * Copyright © 2023 Blue Habit.
 *
 * Unauthorized copying, publishing of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 */

package com.bluehabit.budgetku.component.role.service;

import com.bluehabit.budgetku.common.BaseResponse;
import com.bluehabit.budgetku.common.Constant;
import com.bluehabit.budgetku.common.PermissionUtils;
import com.bluehabit.budgetku.component.role.entity.Permission;
import com.bluehabit.budgetku.component.role.entity.PermissionGroup;
import com.bluehabit.budgetku.component.role.model.DeleteRolesRequest;
import com.bluehabit.budgetku.exception.GeneralErrorException;
import com.bluehabit.budgetku.exception.UnAuthorizedException;
import com.bluehabit.budgetku.component.role.model.CreateRoleRequest;
import com.bluehabit.budgetku.component.role.repo.PermissionGroupRepository;
import com.bluehabit.budgetku.component.role.repo.PermissionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
public class RoleService {
    @Autowired
    private PermissionRepository permissionRepository;

    @Autowired
    private PermissionGroupRepository permissionGroupRepository;

    public ResponseEntity<BaseResponse<Page<Permission>>> getPermissions(Pageable pageable){
        var hasAccess = PermissionUtils.hasAccess(Constant.READ_ROLE);
        if(!hasAccess){
           throw new UnAuthorizedException(1008,"User does not have access!");
        }

        var findAllPermission = permissionRepository.findAll(pageable);

        return BaseResponse.success("Success",findAllPermission);
    }

    public ResponseEntity<BaseResponse<Page<PermissionGroup>>> getRole(Pageable pageable){
        var hasAccess = PermissionUtils.hasAccess(Constant.READ_ROLE);
        if(!hasAccess){
            throw new UnAuthorizedException(1008,"User does not have access!");
        }
        var findAllRole = permissionGroupRepository.findAll(pageable);

        return BaseResponse.success("Success",findAllRole);
    }

    public ResponseEntity<BaseResponse<List<PermissionGroup>>> searchRole(String query){
        var hasAccess = PermissionUtils.hasAccess(Constant.READ_ROLE);
        if(!hasAccess){
            throw new UnAuthorizedException(1008,"User does not have access!");
        }
        var findAllRole = permissionGroupRepository.searchByName(query);

        return BaseResponse.success("Success",findAllRole);
    }

    public ResponseEntity<BaseResponse<PermissionGroup>> createNewRole(CreateRoleRequest createRoleRequest){
        var hasAccess = PermissionUtils.hasAccess(Constant.WRITE_ROLE);
        if(!hasAccess){
            throw  new UnAuthorizedException(1008,"User does not have access!");
        }

        List<Permission> permissions = new ArrayList<>();
        var date = OffsetDateTime.now();
        permissionRepository.findAllById(createRoleRequest.permission()).forEach(permissions::add);
        var permissionGroup = new PermissionGroup(
                UUID.randomUUID().toString(),
                createRoleRequest.roleName(),
                createRoleRequest.roleDescription(),
                permissions,
                date,
                date
        );

        var savedData = permissionGroupRepository.save(permissionGroup);
        return BaseResponse.success("Success create role",savedData);
    }

    public ResponseEntity<BaseResponse<PermissionGroup>> updateNewRole(String roleId,CreateRoleRequest request){
        var hasAccess = PermissionUtils.hasAccess(Constant.WRITE_ROLE);
        if(!hasAccess){
            throw  new UnAuthorizedException(1008,"User does not have access!");
        }

        var findRole = permissionGroupRepository.findById(roleId);
        if(findRole.isEmpty()){
            throw  new GeneralErrorException(400,"Cannot find role!");
        }

        List<Permission> permissions = new ArrayList<>();
        var date = OffsetDateTime.now();
        permissionRepository.findAllById(request.permission()).forEach(permissions::add);

        var role = findRole.get();
        role.setUpdatedAt(date);
        role.setRoleName(request.roleName());
        role.setRoleDescription(request.roleDescription());
        role.setRolePermission(permissions);

        var savedData = permissionGroupRepository.save(role);
        return BaseResponse.success("Success create role",savedData);
    }

    public ResponseEntity<BaseResponse<Object>> deleteRole(String roleId){
        var hasAccess = PermissionUtils.hasAccess(Constant.WRITE_ROLE);
        if(!hasAccess){
            throw new UnAuthorizedException(1008,"User does not have access!");
        }

        var findPermissionGroup = permissionGroupRepository.existsByRoleIdIgnoreCase(roleId);
        if(findPermissionGroup){
            throw new GeneralErrorException(1008,"Role doesn't exist");
        }

        permissionGroupRepository.deleteById(roleId);

        return BaseResponse.success("Success delete role","");
    }

    public  ResponseEntity<BaseResponse<List<String>>> deleteRoles(DeleteRolesRequest request){
        var hasAccess = PermissionUtils.hasAccess(Constant.WRITE_ROLE);
        if(!hasAccess){
            throw new UnAuthorizedException(1008,"User does not have access!");
        }

        permissionGroupRepository.deleteAllById(request.ids());

        return BaseResponse.success("Success delete role",request.ids());
    }

}

File: src/main/kotlin/com/bluehabit/budgetku/data/role/RoleService.kt
/*
 * Copyright © 2023 Blue Habit.
 *
 * Unauthorized copying, publishing of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 */

package com.bluehabit.budgetku.data.role

import com.bluehabit.budgetku.common.Constants.ErrorCode
import com.bluehabit.budgetku.common.Constants.Permission.READ_ROLE
import com.bluehabit.budgetku.common.Constants.Permission.WRITE_ROLE
import com.bluehabit.budgetku.common.exception.DataNotFoundException
import com.bluehabit.budgetku.common.model.BaseResponse
import com.bluehabit.budgetku.common.model.PagingDataResponse
import com.bluehabit.budgetku.common.model.baseResponse
import com.bluehabit.budgetku.common.utils.ValidationUtil
import com.bluehabit.budgetku.common.utils.allowTo
import com.bluehabit.budgetku.common.utils.getTodayDateTimeOffset
import com.bluehabit.budgetku.data.BaseService
import com.bluehabit.budgetku.data.role.permission.PermissionRepository
import com.bluehabit.budgetku.data.role.roleGroup.RoleGroup
import com.bluehabit.budgetku.data.role.roleGroup.RoleGroupRepository
import com.bluehabit.budgetku.data.user.userCredential.UserCredentialRepository
import jakarta.transaction.Transactional
import org.springframework.context.support.ResourceBundleMessageSource
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class RoleService(
    override val i18n: ResourceBundleMessageSource,
    override val userCredentialRepository: UserCredentialRepository,
    override val errorCode: Int = ErrorCode.CODE_ROLE,
    private val roleGroupRepository: RoleGroupRepository,
    private val permissionRepository: PermissionRepository,
    private val validationUtil: ValidationUtil
) : BaseService() {
    @Transactional
    suspend fun getListRoleGroup(
        pageable: Pageable
    ): BaseResponse<PagingDataResponse<RoleGroup>> = buildResponse(
        checkAccess = { it.allowTo(READ_ROLE) }
    ) {

        val findAll = roleGroupRepository.findAll(pageable)
        baseResponse {
            code = HttpStatus.OK.value()
            data = findAll.toResponse()
            message = "Sukses"
        }
    }

    @Transactional
    suspend fun createNewRoleGroup(
        request: RoleRequest
    ): BaseResponse<RoleGroup> = buildResponse(
        checkAccess = { it.allowTo(WRITE_ROLE) }
    ) {
        validationUtil.validate(request)

        val findPermission = permissionRepository.findAllById(request.permissions.orEmpty())

        val date = getTodayDateTimeOffset()


        val roleGroup = RoleGroup(
            roleId = null,
            roleName = request.roleName!!,
            roleDescription = request.roleDescription!!,
            rolePermissions = findPermission.toList(),
            createdAt = date,
            updatedAt = date
        )
        val saveRole = roleGroupRepository.save(roleGroup)

        baseResponse {
            code = HttpStatus.OK.value()
            data = saveRole
            message = "Success"
        }
    }

    @Transactional
    suspend fun updateRoleGroup(
        roleId:String,
        request: RoleUpdateRequest
    ):BaseResponse<RoleGroup> = buildResponse(
        checkAccess = {it.allowTo(WRITE_ROLE)}
    ) {
        validationUtil.validate(request)

        val findRoleById = roleGroupRepository.findByIdOrNull(roleId)
            ?: throw DataNotFoundException(
                translate(""),
                errorDataNotFound
            )
        val findPermission = permissionRepository.findAllById(request.permissions.orEmpty())

        val updateRoleGroup = roleGroupRepository.save(
            findRoleById.copy(
                roleName = request.roleName!!,
                roleDescription = request.roleDescription!!,
                rolePermissions = findPermission.toList(),
                updatedAt = getTodayDateTimeOffset()
            )
        )

        baseResponse {
            code = HttpStatus.OK.value()
            data = updateRoleGroup
            message = "Success"
        }
    }

    @Transactional
    suspend fun deleteRole(
        roleId: String
    ):BaseResponse<RoleGroup> = buildResponse(
        checkAccess = {it.allowTo(WRITE_ROLE)}
    ) {
        val findRole = roleGroupRepository.findByIdOrNull(roleId)
            ?: throw DataNotFoundException(
                translate(""),
                errorDataNotFound
            )
        roleGroupRepository.deleteById(roleId)
        baseResponse {
            code = HttpStatus.OK.value()
            data = findRole
            message = "Success"
        }
    }
}